File: extra_learning/design_patterns/02_factory.py
"""Factory: return a concrete implementation based on a key."""
from __future__ import annotations
from typing import Protocol


class Notifier(Protocol):
    def send(self, msg: str) -> None: ...


class EmailNotifier:
    def send(self, msg: str) -> None: print("email:", msg)


class SMSNotifier:
    def send(self, msg: str) -> None: print("sms:", msg)


_REGISTRY: dict[str, type[Notifier]] = {"email": EmailNotifier, "sms": SMSNotifier}


def make_notifier(kind: str) -> Notifier:
    if kind not in _REGISTRY:
        raise ValueError(f"unknown notifier {kind!r}")
    return _REGISTRY[kind]()


if __name__ == "__main__":
    make_notifier("email").send("hi")
    make_notifier("sms").send("yo")

File: go_for_data_engineering/05_design_patterns/02_factory.go
// Factory: return an interface based on type parameter.
package main

import "fmt"

type Notifier interface {
	Send(msg string) error
}

type emailN struct{}

func (emailN) Send(msg string) error { fmt.Println("email:", msg); return nil }

type smsN struct{}

func (smsN) Send(msg string) error { fmt.Println("sms:", msg); return nil }

func NewNotifier(kind string) (Notifier, error) {
	switch kind {
	case "email":
		return emailN{}, nil
	case "sms":
		return smsN{}, nil
	default:
		return nil, fmt.Errorf("unknown notifier %q", kind)
	}
}

func main() {
	n, _ := NewNotifier("email")
	_ = n.Send("hello")
}
